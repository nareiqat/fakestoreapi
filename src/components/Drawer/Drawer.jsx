//Drawer is a material UI component provides a side panel when a button is clicked


import React from 'react';
import clsx from 'clsx';
import { makeStyles } from '@material-ui/core/styles';
import SwipeableDrawer from '@material-ui/core/SwipeableDrawer';
import Button from '@material-ui/core/Button';
import List from '@material-ui/core/List';
import Divider from '@material-ui/core/Divider';
import FilterByPrice from './Filter/FilterByPrice'
import FilterByCategory from './Filter/FilterByCatergory';
import Sort from './Sort/Sort';

const useStyles = makeStyles({
  list: {
    width: 250,
  },
  fullList: {
    width: 'auto',
  },
  button: {
    border: "1px solid",
    marginLeft:'1rem'
  }
});

export default function SwipeableTemporaryDrawer({resetFilters, handlePriceChange, priceValue, category, handleCategory, sortValue, handleSort }) {
  const classes = useStyles();
  const [state, setState] = React.useState({
    top: false,
    left: false,
    bottom: false,
    right: false,
  });

  const toggleDrawer = (anchor, open) => (event) => {
    if (event && event.type === 'keydown' && (event.key === 'Tab' || event.key === 'Shift')) {
      return;
    }

    setState({ ...state, [anchor]: open });
  };



  const list = (anchor) => (
    <div
      className={clsx(classes.list, {
        [classes.fullList]: anchor === 'top' || anchor === 'bottom',
      })}
      role="presentation"
      onClick={toggleDrawer(anchor, false)}
      onKeyDown={toggleDrawer(anchor, false)}
    >
      <List>
        
        <FilterByPrice handlePriceChange={handlePriceChange} value={priceValue}/>
        <FilterByCategory category={category} handleCategory={handleCategory} />
        <Sort sortValue={sortValue} handleSort={handleSort}/>
        <Button className={classes.button} onClick={resetFilters}>Reset filters</Button>
      </List>
      <Divider />
      
    </div>
  );

  return (
    <div>
      {['left'].map((anchor) => (
        <React.Fragment key={anchor}>
          <Button onClick={toggleDrawer(anchor, true)}>Filters</Button>
          <SwipeableDrawer
            anchor={anchor}
            open={state[anchor]}
            onClose={toggleDrawer(anchor, false)}
            onOpen={toggleDrawer(anchor, true)}
          >
            {list(anchor)}
          </SwipeableDrawer>
        </React.Fragment>
      ))}
    </div>
  );
}