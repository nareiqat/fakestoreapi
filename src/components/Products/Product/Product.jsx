import React from "react";
import {
  Card,
  CardMedia,
  CardContent,
  CardActions,
  Typography,
  IconButton,
} from "@material-ui/core";
import { AddShoppingCart } from "@material-ui/icons";
import useStyles from './styles';
import SimpleModal from "./SimpleModal";

//product object is taken from Products and a card is created with some product information inside it

function Product({ product }) {

  const classes = useStyles();

  // truncate to reduce length of title and make all cards even height, to view full title a description button was added
  function truncate(string,limit){
    return string.length > limit ? `${string.substr(0,limit)}...` : string;
}

  return (
    <Card data-testid ="product-1" className={classes.root}>
      <CardMedia
        className={classes.media}
        image={product.image}
        title={product.title}
      />
      <CardContent>
        <div className={classes.cardContent}>
          <Typography gutterBottom variant="h6" component="h2">
            {truncate(product.title, 30)}
          </Typography>
        </div>
        <div className={classes.cardContent}>
          <Typography gutterBottom variant="h5" component="h2">
            ${product.price}
          </Typography>
        </div>
        
        {/* <SimpleModal product={product} /> */}
       
      </CardContent>
      <CardContent>
      <CardActions disableSpacing className={classes.cardActions}>
        <SimpleModal  product={product} />
        <IconButton aria-label="Add to Cart" >
          <AddShoppingCart />
        </IconButton>
      </CardActions>
      </CardContent>
      
    </Card>
  );
}

export default Product;
