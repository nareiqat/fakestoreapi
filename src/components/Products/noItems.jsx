import React from 'react'

export default function noItems() {
    return (
        <div>
            No products found!
        </div>
    )
}
