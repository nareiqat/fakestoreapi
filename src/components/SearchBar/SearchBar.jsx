import React from "react";
// import { Container, Form, FormGroup,Input, Button} from "reactstrap";
import useStyles from "./styles";
import SearchIcon from "@material-ui/icons/Search";

import { IconButton, TextField} from "@material-ui/core";

function SearchBar({ data, handleSubmit, handleChange }) {
  const classes = useStyles();

  return (
    <div className={classes.main}>
      <form  noValidate autoComplete="off">
        <TextField
          id="standard-basic"
          label="Search for Product"
          onChange={handleChange}
        />
        <IconButton  onClick={handleSubmit}>
          <SearchIcon />
        </IconButton>
      </form>
    </div>
  );
}

export default SearchBar;
