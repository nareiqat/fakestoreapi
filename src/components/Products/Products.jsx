import React from 'react'

import {Grid} from '@material-ui/core'
import Product from './Product/Product';
import useStyles from './Product/styles'

export default function Products({minP, maxP ,data, input, category, value, sortValue, sortAsc, sortDesc}) {
    const classes = useStyles();
 

    return (
        <div className={classes.content}>
        
            <div className={classes.toolbar}/>
                <Grid container justify="center" spacing={4}>
                    {data
                    //sort function, sorts price by hightolow and lowtohigh
                    .sort((a,b) => {
                        if(sortValue === 'hightolow' && a.price > b.price){
                            return -1
                        }else if(sortValue==='lowtohigh' && b.price > a.price){
                            return -1
                        }else if(sortValue === ''){
                            return 0
                        }
                    })
                    .filter((product) => {
                        if(input === null || input === ""){
                            return product
                        } else if(product.title.toLowerCase().includes(input.toLowerCase())){
                            return product
                        }
                    
                    }).filter((product) => {
                        if(category === ""){
                            return product
                        }else if(product.category === category){
                            
                            return product
                        }
                    }).filter((product) => {
                        if(product.price >= minP && product.price <= maxP){
                            return product
                        }else{
                            console.log("no products in the following range")
                        }
                    })
                    .map((product) => (
                        <Grid item key={product.id} xs={12} sm={6} md={4} lg={3}>
                            <Product product={product} />
                        </Grid>
                    ))}
                    
                </Grid>
                
            
        </div>

        
                
            
    )
}

// .filter((product) => {
//     if(sortValue === 'hightolow'){
//         sortDesc(data, product.price)
//     } else if(sortValue==="lowtohigh"){
//         sortAsc(data, price)
//     } else {
//         return product
//     }

// })