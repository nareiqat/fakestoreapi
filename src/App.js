// import logo from './logo.svg';
import './App.css';
import {useState, useEffect} from 'react' 
import Products from './components/Products/Products'
import SearchBar from './components/SearchBar/SearchBar';
function App() {

  const [input, setInput] = useState("")
  const [state, setState] = useState("")
  const [data, setData] = useState([]);
  const [error, setError] = useState('');
  const [loading, setLoading] = useState('');
  const [minPrice, setMinPrice] = useState(0);
  const [maxPrice, setMaxPrice] = useState(1000);
  const apiUrl = 'https://fakestoreapi.com/products'

  useEffect(() => {
    const getData = () => {
      fetch(apiUrl)
        .then((res) => {
          if (res.status >= 400) {
            throw new Error("Can't fetch api data");
          }
          return res.json();
        })
        .then(
          (result) => {
            setData(result)
            setLoading(false);
            console.log(result);
          },
          (err) => {
            setError(err);
            setLoading(false);
            alert(err.message);
            console.log(err.message);
          }
        );
    };
    getData();
  }, [apiUrl]);

  const inputHandler = (event) => {
    setInput(event.target.value);
    // console.log(input);
  };

  const submitHandler = (event) => {
    // console.log(input)
    event.preventDefault()
    
  };

  const filterPrice = (data) => {
    data.forEach(product => {
      if(min)
    })

  }


  return (
    <div>
      <SearchBar handleChange={inputHandler} handleSubmit={submitHandler} input={input}/>
      <Products data={data} input={input}/>
    </div>
  );
}

export default App;
