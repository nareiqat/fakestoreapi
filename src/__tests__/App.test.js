import React from 'react'
import App from '../App'
import toJSON from 'enzyme-to-json';
import Products from '../components/Products/Products';
import Product from '../components/Products/Product/Product';
import Drawer from '../components/Drawer/Drawer'
import {shallow, mount} from "enzyme"
import SearchBar from './../components/SearchBar/SearchBar';
import SimpleModal from '../components/Products/Product/SimpleModal';


const mockProduct = {
  id: 1, 
  category: "men's clothing",
  description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
  image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
  price: 109.95,
  title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
}

const jsonArray = [
  {id: 1, 
  category: "men's clothing",
  description: "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
  image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
  price: 109.95,
  title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
  },
  { 
    category: "men's clothing",
    description: "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
    id: 2,
    image: "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
    price: 22.3,
    title: "Mens Casual Premium Slim Fit T-Shirts "
  }
]

// test if App.js is rendering all components
describe("rendering components in app", () => {
  it("renders App component without crashing", () => {
    shallow(<App />)
  });
})

// test if searchBar UI, products, drawer, product snapshot matches refernce snapshot file stored alongside the test
test("searchbar matches snapshot", () => {
  const searchBar = shallow(<SearchBar />)
  expect(toJSON(searchBar)).toMatchSnapshot();
})

// added some data to mimic some props passed to the products component
test("products matches snapshot", () => {
  const products = shallow(<Products data={jsonArray} product={mockProduct} input={mockProduct.title}/>)
  expect(toJSON(products)).toMatchSnapshot();
})

test("product matches snapshot", () => {
  const product = shallow(<Product product={mockProduct}/>)
  expect(toJSON(product)).toMatchSnapshot();
})

test("drawer matches snapshot", () => {
  const drawer = shallow(<Drawer />)
  expect(toJSON(drawer)).toMatchSnapshot();
})

describe('Test Button component', () => {
  it('Test click event', () => {
    const mockCallBack = jest.fn();

    const simpleModal = shallow((<SimpleModal onClick={mockCallBack} product={mockProduct}/>));
    const button = simpleModal.find("button").simulate('click')
    expect(button.target.value).toEqual(mockProduct.description);
    // console.log(button.value)
  });
});





