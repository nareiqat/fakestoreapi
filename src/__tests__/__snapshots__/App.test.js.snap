// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer matches snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(Button))
    onClick={[Function]}
  >
    Filters
  </WithStyles(ForwardRef(Button))>
  <ForwardRef(SwipeableDrawer)
    anchor="left"
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
  >
    <div
      className="makeStyles-list-7"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="presentation"
    >
      <WithStyles(ForwardRef(List))>
        <FilterByPrice />
        <FilterByCategory />
        <Sort />
      </WithStyles(ForwardRef(List))>
      <WithStyles(ForwardRef(Divider)) />
    </div>
  </ForwardRef(SwipeableDrawer)>
</div>
`;

exports[`product matches snapshot 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-root-3"
  data-testid="product-1"
>
  <WithStyles(ForwardRef(CardMedia))
    className="makeStyles-media-4"
    image="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
    title="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
  />
  <WithStyles(ForwardRef(CardContent))>
    <div
      className="makeStyles-cardContent-6"
    >
      <WithStyles(ForwardRef(Typography))
        component="h2"
        gutterBottom={true}
        variant="h6"
      >
        Fjallraven - Foldsack No. 1 Ba...
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-cardContent-6"
    >
      <WithStyles(ForwardRef(Typography))
        component="h2"
        gutterBottom={true}
        variant="h5"
      >
        $
        109.95
      </WithStyles(ForwardRef(Typography))>
    </div>
  </WithStyles(ForwardRef(CardContent))>
  <WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(CardActions))
      className="makeStyles-cardActions-5"
      disableSpacing={true}
    >
      <SimpleModal
        product={
          Object {
            "category": "men's clothing",
            "description": "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
            "id": 1,
            "image": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
            "price": 109.95,
            "title": "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
          }
        }
      />
      <WithStyles(ForwardRef(IconButton))
        aria-label="Add to Cart"
      >
        <Memo(ForwardRef(AddShoppingCartIcon)) />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(CardActions))>
  </WithStyles(ForwardRef(CardContent))>
</WithStyles(ForwardRef(Card))>
`;

exports[`products matches snapshot 1`] = `
<div>
  <div />
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
    spacing={4}
  />
</div>
`;

exports[`searchbar matches snapshot 1`] = `
<div
  className="makeStyles-main-1"
>
  <form
    autoComplete="off"
    noValidate={true}
  >
    <WithStyles(ForwardRef(TextField))
      id="standard-basic"
      label="Search for Product"
    />
    <WithStyles(ForwardRef(IconButton))>
      <Memo(ForwardRef(SearchIcon)) />
    </WithStyles(ForwardRef(IconButton))>
  </form>
</div>
`;
